src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreDataSource.cs

```diff
@@ -206,5 +222,67 @@ public Task UnloadAsync()
 
             return Task.CompletedTask;
         }
+
+        // NuGet calls this method to wait project to nominate restoring
+        // If the project has no pending restore data, it will return a completed task.
+        // Otherwise a task which will be completed once the project norminate the next restore
+        // the task will be cancelled, if the project system decide it no longer need restore (for example: the restore state has no change)
+        // the task will be failed, if the project system runs into a problem, so it cannot get correct data to norminate a restore (DT build failed)
+        public Task WhenNominated(CancellationToken cancellationToken)
+        {
+            if (_hasPendingNomination == false || cancellationToken.IsCancellationRequested)
+            {
+                return Task.CompletedTask;
+            }
+
+            JoinableTask joinableTask = JoinableFactory.RunAsync(() =>
+            {
+                while (true)
+                {
+                    if (_nominateNextRestore == true || cancellationToken.IsCancellationRequested)
+                    {
+                        return Task.CompletedTask;
+                    }
+                }
+            });
+
+            return joinableTask.Task;
```

Comment:
Wouldn't this cause a deadlock when called by NuGet?

Say you have a solution with 2 projects.
1 has nominated and for whatever reason NuGet calls WhenNominated on the one that hasn't nominated, this task would only return when the NominateProjectAsync task completes, which in turn would only complete when NuGet actually runs restore, but NuGet can't run restore because it's blocked on when nominated. 

I'm wondering if we need to rethink this part of the API :/

Reply:
Yeah, you are right. I think I left quite some comments 8 days ago, and many of them were not yet addressed. I think we need make sure some essential things are correct:

1, projects need be registered before first nomination from other project is called.

2, HasPendingNormination should return false and earlier task from WhenNorminated must be completed right after the project starts the normination, but should not after the normination is done. (Otherwise, obviously, we are creating a system wide dead lock). So the code must split starting normination, and waiting on it to two steps.

3, we should never have code to spin CPU to wait a status change, it should return a task.

Reply:
logic implemented in latest commit